libraries/core_libs/consensus/src/pillar_chain/pillar_chain_manager.cpp

```diff
@@ -0,0 +1,489 @@
+                     << ", period: " << current_pillar_block_->getPeriod() << ". Current period " << block_num;
+        return nullptr;
+      }
+
+      // Save current pillar block and 2t+1 votes into db
+      if (!pushPillarBlock(PillarBlockData{current_pillar_block_, std::move(two_t_plus_one_votes)})) {
+        // This should never happen
+        LOG(log_er_) << "Unable to push pillar block: " << current_pillar_block_->getHash() << ", period "
+                     << current_pillar_block_->getPeriod();
+        assert(false);
+        return nullptr;
+      }
+    }
+
+    // This should never happen
+    if (current_pillar_block_vote_counts_.empty()) {
+      LOG(log_er_) << "Empty current pillar block vote counts, new pillar block period " << block_num;
+      assert(false);
+      return nullptr;
+    }
+
+    previous_pillar_block_hash = current_pillar_block_->getHash();
+
+    // Get validators vote counts changes between the current and previous pillar block
+    votes_count_changes = getOrderedValidatorsVoteCountsChanges(new_vote_counts, current_pillar_block_vote_counts_);
+  }
+
+  // TODO: provide bridge root ???
+  const auto pillar_block = std::make_shared<PillarBlock>(block_num, block_data->final_chain_blk->state_root, h256{},
+                                                          std::move(votes_count_changes), previous_pillar_block_hash);
+
+  // Check if some pillar block was not skipped
+  if (!isValidPillarBlock(pillar_block)) {
+    LOG(log_er_) << "Newly created pillar block " << pillar_block->getHash() << "with period "
+                 << pillar_block->getPeriod() << " is invalid";
+    return nullptr;
+  }
+
+  LOG(log_nf_) << "New pillar block " << pillar_block->getHash() << " with period " << pillar_block->getPeriod()
+               << " created";
+
+  {
+    std::scoped_lock<std::shared_mutex> lock(mutex_);
+    current_pillar_block_ = pillar_block;
+    current_pillar_block_vote_counts_ = std::move(new_vote_counts);
+
+    auto batch = db_->createWriteBatch();
+    db_->saveCurrentPillarBlock(current_pillar_block_, batch);
+    db_->saveCurrentPillarBlockVoteCounts(current_pillar_block_vote_counts_, batch);
+
+    // Commit DB
+    db_->commitWriteBatch(batch);
+  }
+
+  return pillar_block;
+}
+
+bool PillarChainManager::genAndPlacePillarVote(const PillarBlock::Hash& pillar_block_hash, const secret_t& node_sk,
+                                               bool is_pbft_syncing) {
+  std::shared_ptr<PillarVote> vote;
+  {
+    std::shared_lock<std::shared_mutex> lock(mutex_);
+    if (!current_pillar_block_) {
+      // This should never happen
+      LOG(log_er_) << "Unable to gen pillar vote. No pillar block present, pillar_block_hash " << pillar_block_hash;
+      assert(false);
+      return false;
+    }
+
+    if (!final_chain_->dpos_is_eligible(current_pillar_block_->getPeriod(), toAddress(node_sk))) {
```

Comment:
fixed, for the try catch we dont need. I added the comment why:
// No need to catch ErrFutureBlock because pillar block as well as pillar vote are created only after the pbft block
// with current_pillar_block->getPeriod() was finalized